src/cargo/core/resolver/dep_cache.rs

```diff
@@ -164,26 +165,58 @@ impl<'a> RegistryQueryer<'a> {
             }
         }
 
+        // Is this an exact req of the form `=1.0.0` or `=1.0.0+metadata`. We
+        // boost priority of exact matches to such requirements, taking into
+        // account build metadata, which is otherwise not considered relevant in
+        // whether a version matches a requirement.
+        let is_exact_req = match dep.version_req() {
+            OptVersionReq::Any => false,
+            OptVersionReq::Req(req) => {
+                dep.specified_req().is_some()
+                    && req.comparators.len() == 1
+                    && req.comparators[0].op == semver::Op::Exact
+                    && req.comparators[0].patch.is_some()
+            }
+        };
+
+        let prioritize_exact_match = |v: &semver::Version| -> bool {
+            dep.specified_req()
+                .unwrap()
+                .trim()
+                .trim_end_matches(&v.to_string())
+                .trim_end()
+                == "="
+        };
+
         // When we attempt versions for a package we'll want to do so in a
-        // sorted fashion to pick the "best candidates" first. Currently we try
-        // prioritized summaries (those in `try_to_use`) and failing that we
-        // list everything from the maximum version to the lowest version.
+        // sorted fashion to pick the "best candidates" first.
         ret.sort_unstable_by(|a, b| {
+            if is_exact_req {
+                // Prefer a version that precisely matches an `=` requirement.
+                let a_exact_match = prioritize_exact_match(a.version());
+                let b_exact_match = prioritize_exact_match(b.version());
+                let exact_match_cmp = a_exact_match.cmp(&b_exact_match).reverse();
+                if exact_match_cmp != Ordering::Equal {
+                    return exact_match_cmp;
+                }
+            }
+
+            // Prefer prioritized summaries (those in `try_to_use`).
             let a_in_previous = self.try_to_use.contains(&a.package_id());
```

Comment:
How come this is no longer prioritized? Were you running into issues which forced the new comparison to come first?

Reply:
Naively I'd sort of expect this to come after the semver compare, in that if we really care about same-version-different-build-metadata that seems like it'd be part of the semver comparison function below for ordering, basically doing first a sort on version and then a sort on "does this equal or not"

Reply:
Good call, I think the previous_cmp comparison should go above the exact_match_cmp comparison. The case I had in mind writing this code was that somebody has `=x.y.z+build1` in Cargo.toml, resolves it, then edits Cargo.toml to say `=x.y.z+build2`, and resolves again. In that case probably what they meant is for the exact match (build2) to be chosen.

However, that person's use case is less important than somebody else who has `=x.y.z+build2` in Cargo.toml, ran `cargo update --precise x.y.z+build1`, then makes some unrelated change (maybe adding a dependency) in Cargo.toml which causes the resolver to rerun. That person probably wants the previous match (build1) to remain chosen.

Reply:
Yeah I think this is where multiple versions with the same build metadata doesn't really make sense. I see what you mean about how one ordering here respects lock files while the other ordering respects `--precise`. I don't think we can support both simultaneously because we don't know what manifest was used to generate a lock file. It seems reasonable to prefer the `--precise` flag for now I guess? 

This whole area seems like it's just full of surprising gotchas and is ripe for us receiving lots of bug reports if it's used too much... Hopefully it's not though?